discretize/tree.cpp

```diff
@@ -0,0 +1,1166 @@
+#include <vector>
+#include <map>
+#include "tree.h"
+#include <iostream>
+
+Node::Node(){
+    location_ind[0] = 0;
+    location_ind[1] = 0;
+    location_ind[2] = 0;
+    location[0] = 0;
+    location[1] = 0;
+    location[2] = 0;
+    key = 0;
+    reference = 0;
+    index = 0;
+    hanging = false;
+    parents[0] = NULL;
+    parents[1] = NULL;
+    parents[2] = NULL;
+    parents[3] = NULL;
+};
+
+Node::Node(int_t ix, int_t iy, int_t iz, double* xs, double *ys, double *zs){
+    location_ind[0] = ix;
+    location_ind[1] = iy;
+    location_ind[2] = iz;
+    location[0] = xs[ix];
+    location[1] = ys[iy];
+    location[2] = zs[iz];
+    key = key_func(ix, iy, iz);
+    reference = 0;
+    index = 0;
+    hanging = false;
+    parents[0] = NULL;
+    parents[1] = NULL;
+    parents[2] = NULL;
+    parents[3] = NULL;
+};
+
+Edge::Edge(){
+    location_ind[0] = 0;
+    location_ind[1] = 0;
+    location_ind[2] = 0;
+    location[0] = 0;
+    location[1] = 0;
+    location[2] = 0;
+    key = 0;
+    index = 0;
+    reference = 0;
+    length = 0.0;
+    hanging = false;
+    points[0] = NULL;
+    points[1] = NULL;
+    parents[0] = NULL;
+    parents[1] = NULL;
+};
+
+Edge::Edge(Node& p1, Node& p2){
+      points[0] = &p1;
+      points[1] = &p2;
+      int_t ix, iy, iz;
+      ix = (p1.location_ind[0]+p2.location_ind[0])/2;
+      iy = (p1.location_ind[1]+p2.location_ind[1])/2;
+      iz = (p1.location_ind[2]+p2.location_ind[2])/2;
+      key = key_func(ix, iy, iz);
+      location_ind[0] = ix;
+      location_ind[1] = iy;
+      location_ind[2] = iz;
+
+      location[0] = (p1[0]+p2[0]) * 0.5;
+      location[1] = (p1[1]+p2[1]) * 0.5;
+      location[2] = (p1[2]+p2[2]) * 0.5;
+      length = (p2[0]-p1[0])
```

Comment:
This is only true if the edges are axes-aligned: currently, this is the case, but down-the-road we can look towards non-axes aligned edges (eg: curvi mesh). Might be worth just adding a comment in the code here